Makefile

```
@@ -27,32 +27,46 @@ escape_double_quote = $(subst $\",$\\",$(1))
 # Utility function to require a variable to be defined
 require_defined = $(if $(value $1),$(value $1),$(error "$1 needs to be defined"))
 
+# What implementation to build
+VARIANT ?= trivial
 # Where to put build files
 OUT ?= out
 # Where to put typescript files
 TS_OUT ?= typescript
 
 .PHONY: all clean test typescript
 
-all: \
-	$(OUT)/baseDAO.tz
+all: $(OUT)/trivialDAO.tz $(OUT)/registryDAO.tz $(OUT)/treasuryDAO.tz $(OUT)/testDAO.tz
+
+trivialDAO: VARIANT = trivial
+trivialDAO: baseDAO
+
+registryDAO: VARIANT = registry
+registryDAO: baseDAO
+
+treasuryDAO: VARIANT = treasury
+treasuryDAO: baseDAO
+
+testDAO: VARIANT = test
+testDAO: baseDAO
 
 # Compile LIGO contract into its michelson representation.
-$(OUT)/baseDAO.tz: src/**
+$(OUT)/$(VARIANT)DAO.tz: src/**
+	cp src/$(VARIANT)DAO.mligo src/implementation.mligo
 	mkdir -p $(OUT)
 	# ============== Compiling contract ============== #
-	$(BUILD) src/base_DAO.mligo base_DAO_contract --output-file $(OUT)/baseDAO.tz
+	$(BUILD) src/base_DAO.mligo -e base_DAO_contract --output-file $(OUT)/$(VARIANT)DAO.tz
```

while we are at it, this could use another automatic variable, `$@`:
```suggestion
	$(BUILD) src/base_DAO.mligo -e base_DAO_contract --output-file $@
```